
<br>
<h5> Similar cu ce am facut la tutoriat: Problema diamantului, Upcasting, Downcasting, Polimorfism la executie</h5>
<code>

    class User {
    protected:
    std::string name;
    int age;
    public:
    User(const std::string & n, int a) : name(n), age(a) {}
    
        virtual void showAge() const {
            std::cout << "Userul nu numele= "<< name << " are varsta "<< age << '\n';
        }
        void showAge(int) const {
            std::cout << "Supraincarcarea fct showAge din User"<< '\n';
        }
    };
    
    class Student: public User {
    protected:
    int studentId;
    float average;
    public:
    Student(const std::string &n, int a, int id, float avg) : User(n, a), studentId(id), average(avg) {}
    
        void showAvg() const {
            std::cout << "Studentul cu id = "<< studentId << " are media "<< average << '\n';
    
    }
    void showAge() const override{
    std::cout << "Studentul nu numele= "<< name << " are varsta "<< age << '\n';
    }
    };
    
    class Teacher: public User {
    protected:
    int teacherId;
    int courseId;
    public:
    Teacher(const std::string &n, int a, int id, int course) : User(n, a), teacherId(id), courseId(course) {}
    
        void showCourseId() const {
            std::cout << "Profesorul cu id = "<< teacherId << " are cursul cu id =  "<< courseId << '\n';
    
        }
        void showAge() const override{
            std::cout << "Profesorul nu numele= "<< name << " are varsta "<< age << '\n';
        }
    };
    
    // problema diamantului
    class A {
    public:
        int x;
        int y;
    };
    class B:public virtual A {
    
    };
    
    class C:public virtual A {
    
    };
    
    class D:public B, public C {
    public:
    void showXY(){
        std::cout << x << " "<< y <<'\n'; // apare ambiguitate daca omitem cuvantul virtual de la mostenirile B:A si C:A
    }
    };
    
    int main() {
        D* d= new D();
        d->showXY();
    
        User* user = new Student("Ana", 17, 2897, 9.88);
        user->showAge();
        user->showAge(5);

        // down cast -> decomenteaza
        //    Student* student = dynamic_cast<Student*>(user);
        //    if (student != nullptr) {
        //        student->showAvg();
        //    }
    
        Student* student = new Student("Mara", 23, 34197, 8.09);
        student->showAge();
        student->User::showAge();
    
        std::vector<User*> useri;
        useri.push_back(user);
    
        User* user2 = new Teacher("Maria", 36, 89, 99);
        useri.push_back(user2);
        std::cout<<"Afisare varste useri:\n";
    
        for(auto user: useri) {
            //utilizare typeid din <<typeinfo> -> trebuie sa includeti : #include <typeinfo>
            //https://en.cppreference.com/w/cpp/types/type_info/name
            std::cout << "Tip Pointer : " << typeid(user).name() << "; Tip obiect pointat :" << typeid(*user).name() << " ";
            user->showAge();
        }
    
    }


</code>

<br>

<h5> Despre destructorul virtual pur :</h5>

<code>

    //DESTRUCTORUL POATE FI VIRTUAL (E SI RECOMANDAT SA FIE)
    // Mai mult decat atat, poate fi si virtual pur -> dar trebuie sa ii scriem corpul in clasa de baza (in clasa in care e virtual pur)

    class Person {
    public:
        virtual ~Person() = 0; // destructor virtual pur -> clasa e automat ABSTRACTA = nu poate fi instantiata
    };
    
    Person:: ~Person() {
        // desi destructorul e virtual -> in acest caz trebuie sa ii scriem corpul in clasa de baza !!!, altfel partea de baza nu poate fi stearsa
        // chiar daca acum destructorul are corp, el ramane functie virtuala pura si trebuie implementat in clasele derivate
        cout <<"Destructor virtual pur Person\n";
    }
    
    
    class Employee: public Person {
    public:
        // pentru ca in clasa de baza exista o functie virtuala pura,
        // daca noi mostenim aceasta clasa (Person) trebuie neaparat sa scriem functia virtuala pura (altfel si Employee ramane abstracta)
        ~Employee() override {
            cout <<"Destructor Employee\n";
        }
    };

    int main() {
        Person* person = new Employee();
        delete person;
        return 0;
    }
</code>

<br>
<h5> Mostenire din clasa std::exception, pentru a ne crea exceptia noastra : (exemplu luat din curs) </h5>

<code>
    
    Mai intai m-am uitat in documentatia C++ pentru a vedea cum arata clasa std:: exception ( daca puneti using namespace std; nu mai tb pus std::):
    //https://cplusplus.com/reference/exception/exception/

    class exception {
    public:
        exception () noexcept; // constructor 
        exception (const exception&) noexcept; // constructor copiere
        exception& operator= (const exception&) noexcept; // operator atribuire
        virtual ~exception(); // destructor virtual (pentru ca nu e virtual pur nu trebuie rescris in clasa derivata neaparat)
        virtual const char* what() const noexcept; // functia what de fapt returneaza mesajul erorii -> pe ea tb sa o rescriem in clasa derivata ( vedem ca e si virtuala, deci perfect!)
    }

</code>


<code>

    //EXEMPLU preluat din curs OOP

    using namespace std;
    class MyException: public std::exception {
    public:
    const char* what() const throw () { // suprascriere functie care in std::exception e VIRTUALA
    return "Exceptia mea\n";
    }
    };
    
    class MathException: public MyException {
    public:
    const char* what() const throw () {
    return "Exceptia mea matematica\n";
    }
    };
    
    int main() {
        // in general, e bine sa prindem in catch exceptiile in ordinea particular general
        //adica intai exceptiile noastre, si apoi urcam in ierarhie pana la clasa std::exception
        /// Luam primul exemplu:
        //In exemplul acesta prindem GRESIT: mai intai o exceptie mai generala (MyException) si apoi una mai particulara (MathException, care deriva din MyException)
        /// Mesajul afisat va fi -> " Exceptia mea " == mesajul din MyException ( cum ne si asteptam, clasa de baza poate lua locul derivatei si va intra pe primul catch)
        try {
            throw MathException();
            } catch (MyException e) {
                std::cout << e.what() << "\n";
            } catch (MathException e) {
                std::cout << e.what() << "\n";
            }
        ////////

        //Daca schimbam doar putin exemplul si prindem exceptiile cu tipul referinta
        //Ne-am astepta sa intre tot pe prima ramura
        // Dar mesajul afisat vedem ca e " Exceptia mea matematica " == Mesajul din clasa MathException (care e al doilea catch)
        // De ce?
        // Deoarece are loc fenomenul despre care am vorbit la tutoriat: polimorfism
        // Cand avem pointeri si referinte, o ierarhie, o functie virtuala in clasa de baza ( aici e vb de fct what() ) si upcasting 
        // atunci apelul la functia respectiva prin pointer sau referinta va apela functia din clasa reala a obiectului
        // ATENTIE MAI SUS VEDEM CA APELUL la functie prin obiect nu beneficiaza de beneficiile acestea ( se apeleaza functia din clasa careia e obiectului)
        //Concluzie -> se intra tot pe prima ramura dar fiind vorba de referinta, apelul la functie se face din clasa reala a obiectului (polimorfism la executie)
        try {
            throw MathException();
            } catch (MyException &e) {
                std::cout << e.what() << "\n";
            } catch (MathException &e) {
                std::cout << e.what() << "\n";
            }

        // Putem testa cu mai multe cout-uri pentru a vedea mai clar ce se intampla:

        try {
            std:: cout << "In try, arunc o exceptie de tip MathException"<<'\n';
            throw MathException();
        } catch (MyException &e) {
            std:: cout << "Prima ramura de catch-> cea care prinde MyException& e" << '\n';
            std:: cout << "Tipul REAL al exceptiei prinse :"<< typeid(e).name() << '\n';
            std::cout << e.what() << "\n";
        } catch (MathException &e) {
            std:: cout << "A doua ramura de catch -> cea care prinde MathException& e" << '\n';
            std:: cout << "Tipul REAL al exceptiei prinse :"<< typeid(e).name() << '\n';
            std::cout << e.what() << "\n";
        }
        // Vedem ca typeid().name() -> returneaza un sir de caractere care contine litera 'P' daca este vb de un pointer 
        // apoi un numar care reprezinta cate caractere are numele clasei si apoi numele clasei la executie 

        return 0;
    }
</code>